### PR TITLE
feat: datatable added totalfooter

### PR DESCRIPTION
**Pull request by ShubhamBhagat-p44**
Closes # 

## 📝 Description

Added Total footer row to dataTable component

## Screenshots

image

## Merge checklist

- [ ] Added/updated tests
- [ ] Added changeset

### CODE REVIEW

File: packages/react/src/components/DataTable/DataTable.types.ts [hunk 1]
```diff
@@ -208,3 +228,21 @@ export interface DataTableProps<TData extends RowData> {
...
+  totalsKeyword: string;
+}
+
+export type TotalsDataObj = Record<string, TotalsHeaderObj>;
+
+export interface TotalsHeaderObj {
+  value: number;
+  has_custom_html: boolean;
+  html: string;
+  links: Link[];
```
Comment (vitaliy-p44) [inline packages/react/src/components/DataTable/DataTable.types.ts:243]: what do these props do?

Comment (ShubhamBhagat-p44) [reply packages/react/src/components/DataTable/DataTable.types.ts:243]: These will provide user option to add links and html to total column

Comment (vitaliy-p44) [reply packages/react/src/components/DataTable/DataTable.types.ts:243]: I don't see it used anywhere

Comment (ShubhamBhagat-p44) [reply packages/react/src/components/DataTable/DataTable.types.ts]: @vitaliy-p44 I have updated these prop to use generic type

**Comment by vitaliy-p44** [comment]
It still looks like an overbuilt and I don't see any sense adding this to manifest. 
Probably we can just support passing footer row data? Would love to see other people thought
@dmorton-p44 @jklegsethp44 @p44dyoachim

**Comment by jklegsethp44** [comment]
> It still looks like an overbuilt and I don't see any sense adding this to manifest. Probably we can just support passing footer row data? Would love to see other people thoughts @dmorton-p44 @jklegsethp44 @p44dyoachim

I tend to agree. Is this something for your app only or have others been requesting it? Would what @vitaliy-p44 suggests not be enough for what you need?

> Probably we can just support passing footer row data?

**Comment by dmorton-p44** [comment]
I tend to lean towards what @vitaliy-p44 is suggesting - 

However if we are wanting to introduce summary rows as a common thing across our design system, I am not totally opposed to this. 

This component is already a tightly coupled implementation of https://tanstack.com/table/v8, and its intention is to enforce some opinions on how we render table data

I think if a dev needs flexibility outside of what this standard component suggests, they would resort to using the existing `Table` `TableRow` `TableCell` type components to compose something.

**Comment by ShubhamBhagat-p44** [comment]
> It still looks like an overbuilt and I don't see any sense adding this to manifest. Probably we can just support passing footer row data? Would love to see other people thoughts @dmorton-p44 @jklegsethp44 @p44dyoachim

Isn't it the same thing just instead of footer row data prop is named as totals row data, but i guess both do the same thing

**Comment by ShubhamBhagat-p44** [comment]
> > It still looks like an overbuilt and I don't see any sense adding this to manifest. Probably we can just support passing footer row data? Would love to see other people thoughts @dmorton-p44 @jklegsethp44 @p44dyoachim
> 
> I tend to agree. Is this something for your app only or have others been requesting it? Would what @vitaliy-p44 suggests not be enough for what you need?
> 
> > Probably we can just support passing footer row data?

Yes @jklegsethp44 for now these is something required specifically for  our app but as per my understanding from discussion i had with design team soon these is going to be a requirement in most of the tables used. either way these is controlled via an optional prop .

**Comment by ShubhamBhagat-p44** [comment]
> I tend to lean towards what @vitaliy-p44 is suggesting -
> 
> However if we are wanting to introduce summary rows as a common thing across our design system, I am not totally opposed to this. It would make the question of implementing this on both design / engineering trivial.
> 
> This component is already a tightly coupled implementation of https://tanstack.com/table/v8, and its intention is to enforce some opinions on how we render table data
> 
> I think if a dev needs flexibility outside of what this standard component suggests, they would resort to using the existing `Table` `TableRow` `TableCell` type components to compose something.

Yes @dmorton-p44 i understand thats why i have referred these [example](https://tanstack.com/table/v8/docs/examples/react/column-dnd) already provided by tanstack .

File: packages/react/stories/DataTable.stories.tsx [hunk 2]
```diff
@@ -394,3 +403,180 @@ export const Loading = () => {
...
+      html: '0.3 p.p.',
+      links: [
+        {
+          url: false,
+        },
+      ],
+    },
+    birthday: {
+      value: 0.005304,
+      has_custom_html: false,
```
Comment (dmorton-p44) [inline packages/react/stories/DataTable.stories.tsx:545]: use camel case

File: packages/react/stories/DataTable.stories.tsx [hunk 3]
```diff
@@ -33,34 +34,42 @@ export const Default = () => {
...
     },
     {
       header: 'City',
       accessorKey: 'city',
+      footer: (props) => props.column.id,
     },
     {
       header: 'Subscribed',
       accessorKey: 'isSubscribed',
+      footer: (props) => props.column.id,
```
Comment (vitaliy-p44) [inline packages/react/stories/DataTable.stories.tsx:67]: I don't see `footer` props is used anywhere

File: packages/react/stories/DataTable.stories.tsx [hunk 4]
```diff
@@ -394,3 +403,180 @@ export const Loading = () => {
...
+    city: faker.address.city(),
+    isSubscribed: faker.datatype.boolean(),
+    birthday: faker.date.past().toLocaleDateString('en-US'),
+  }));
+
+  const columns: DataTableColumnDef<Person>[] = [
+    {
+      header: 'First Name',
+      accessorKey: 'firstName',
+      footer: (props) => props.column.id,
```
Comment (vitaliy-p44) [inline packages/react/stories/DataTable.stories.tsx:433]: and here

**Review by vitaliy-p44 — COMMENTED** [review]
Let's clean it up to be useful and easy to understand.
I would suggest to treat all of these just as "footer" instead of "totals footer"

**Review by dmorton-p44 — COMMENTED** [review]
I am for making this change if needed - some minor comments

File: packages/react/src/components/DataTable/components/DataTableFooter/DataTableFooter.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      })}
+    >
+      {table.getFooterGroups().map((footerGroup) => (
+        <tr key={footerGroup.id}>
+          {footerGroup.headers.map((header, i) => (
+            <DataTableFooterColumn
+              key={header.id}
+              getTotalValue={getTotalValue}
+              header={header}
+              index={i}
```
[the remark this answers is not in the document]
Comment (ShubhamBhagat-p44) [reply packages/react/src/components/DataTable/components/DataTableFooter/DataTableFooter.tsx:22]: these is not for key we are using header.id for key , index is provided as prop so the user will have flexibility to place summary keyword on different column for ex: 1st column or last column

File: packages/react/stories/DataTable.stories.tsx (hunk 3 above)
Comment (ShubhamBhagat-p44) [reply packages/react/stories/DataTable.stories.tsx:67]: removing these

File: packages/react/stories/DataTable.stories.tsx (hunk 2 above)
Comment (ShubhamBhagat-p44) [reply packages/react/stories/DataTable.stories.tsx:545]: these is sample data user can pass

File: packages/react/src/components/DataTable/components/DataTableFooterColumn/DataTableFooterColumn.tsx [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    <th
+      key={header.id}
+      className={cx('manifest-table__cell', 'manifest-table__cell--header', {
+        'manifest-table__cell--sticky-header': table.options.enableStickyHeader,
+        'manifest-table__cell--pinned': '',
+      })}
+      colSpan={header.colSpan}
+      style={{ ...styles }}
+    >
+      {!Object.prototype.hasOwnProperty.call(totalsData, header.id) || index === 0
```
[the remark this answers is not in the document]
Comment (ShubhamBhagat-p44) [reply packages/react/src/components/DataTable/components/DataTableFooterColumn/DataTableFooterColumn.tsx:26]: yeah sure

File: packages/react/stories/DataTable.stories.tsx (hunk 4 above)
Comment (ShubhamBhagat-p44) [reply packages/react/stories/DataTable.stories.tsx]: removing these also

File: packages/react/stories/DataTable.stories.tsx [hunk 7]
```diff
@@ -394,3 +403,175 @@ export const Loading = () => {
...
+      html: '0.3 p.p.',
+      links: [
+        {
+          url: false,
+        },
+      ],
+    },
+    birthday: {
+      value: 0.005304,
+      hasCustomHtml: false,
```
Comment (vitaliy-p44) [inline packages/react/stories/DataTable.stories.tsx:537]: I see that only **value** prop is used

File: packages/react/stories/DataTable.stories.tsx (hunk 3 above)
Comment (vitaliy-p44) [reply packages/react/stories/DataTable.stories.tsx:67]: they're still there

File: .changeset/dull-grapes-raise.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'@project44-manifest/react': minor
+---
+
+Added support for totsl footer row in datatable
```
Comment (vitaliy-p44) [inline .changeset/dull-grapes-raise.md:5]: totsl

File: packages/react/stories/DataTable.stories.tsx (hunk 7 above)
Comment (ShubhamBhagat-p44) [reply packages/react/stories/DataTable.stories.tsx]: My bad removing these

File: .changeset/dull-grapes-raise.md (hunk 8 above)
Comment (ShubhamBhagat-p44) [reply .changeset/dull-grapes-raise.md]: Fixed these

**Review by vitaliy-p44 — APPROVED** [review]
I think it's good enough to merge to not block other work. though still may require some further polishing.

**Pull request merged**